// easynodeapi/NodeAPI.js
var fs = require("fs");
var qs = require("querystring");
var types = require("./types.js");

console.log("Start API");
module.exports = function (req,res) {
    ServerFunc = [];
    return {
    res:res,
    req:req,
    View:function(file)
    {
        fs.readFile(file, function (error, data) {
            if (error) {
                res.write("Error ",error);
                res.end();
            }
            else {
                res.writeHead(200, { 'Content-Type': 'text/html' });
                res.write(data);
                res.end();
            }
        })
    },
    File:function(file)
    {
        var roz = file.substring(file.indexOf(".",1)+1,file.length);
        fs.readFile(file, function (error, data) {
            if (error) {
                res.write("Error " + error.message);
                res.end();
            }
            else {
                var actType = types.find(function(element){
                    return element.type == roz
                })
                res.writeHead(200,{'Content-Type' : actType ? actType.ctype : "text/plain"});
                res.write(data);
                res.end();
            }
        })
    },
    Json:function(data)
    {
        res.writeHead(200, { 'Content-Type': "application/json" })
        res.write(JSON.stringify(data));
        res.end();
    },
    SendPOSTData:function(func)
    {
        var str = "";
        var obj = {};
        var dat = new Date();
        req.on("data", function (data) {
            str += data;
        })
        req.on("end", function (data) {
            obj = qs.parse(str);
            if (func)
            {
                func(obj);
            }
            res.end(JSON.stringify(obj));
        })
    },
    GetPOSTData:function(func){
        var str = "";
        var obj = {};
        var dat = new Date();
        req.on("data", function (data) {
            str += data;
        })
        req.on("end", function (data) {
            obj = qs.parse(str);
            if (func)
            {
                func(obj)
            }
        })
    },
    SavePOSTData: function (name,func) {
        var str = "";
        var obj = {};
        var dat = new Date();
        req.on("data", function (data) {
            str += data;
        })
        req.on("end", function (data) {
            obj = qs.parse(str);
            if (func)
            {
                func(obj)
            }
            this.Data[name] = obj;
        })
    },
    OnGETUrl: function (url, func) {
        var reqUrl = UrlFilter(url);
        ServerFunc.push({url:reqUrl,method:"GET",func:func});
    },
    OnPOSTUrl: function ( url, func) {
        var reqUrl = UrlFilter(url);
        ServerFunc.push({url:reqUrl,method:"POST",func:func});
    },
    OnUrl:function(method,url,func){
        var reqUrl = UrlFilter(url);
        ServerFunc.push({url:reqUrl,method:method,func:func});
    },
    Default:function(func){
        this.ServerDefault = func;
    },
    Tab:{
        RemFromArray:function (index,table)
        {
            if(index = table.length)
            {
                table.pop()
                return table;
            }
            else
            {
                for(var i=index;i<table.length-1;i++)
                {
                    table[index] = table[index + 1];
                }
                table.pop();
                return table;
            }
        }
    },
    executedReq:false,
    ServerDefault:null,
    Server:function(){
        var req = UrlFilter(this.req.url);
        if(!this.ServerDefault)
        {
            console.log("You must set the default response");
        }
        for(var i=0;i<ServerFunc.length;i++)
        {
            if (req == ServerFunc[i].url && this.req.method == ServerFunc[i].method)
            {
                if(this.req.method == "GET")
                {
                    ServerFunc[i].func(UrlGetFilter(this.req.url));
                }
                else if(this.req.method == "POST")
                {
                    this.GetPOSTData(function(data){
                        ServerFunc[i].func(data);
                    })
                }
                else
                {
                    ServerFunc[i].func();
                }
                this.executedReq = true;
                return;
            }
        }
        if(!this.executedReq)
        {
            this.ServerDefault();
        }
    }
    }
}
function AddToObject(obj,name,value)
{
    eval("obj." + name + " = " + JSON.stringify(value));
    return obj;
}
function UrlFilter(url)
{
    var ret = "";
    for(var i=0;i<url.length;i++)
    {
        if(url[i] == "?")
        {
            return ret;
        }
        else
        {
            ret += url[i];
        }
    }
    return url;
}
function UrlGetFilter(url)
{
    var obj = {};
    var ret = "";
    var beginGet = false;
    for(var i=0;i<url.length;i++)
    {
        if(url[i] == "?")
        {
            beginGet = true;
        }
        else if(beginGet)
        {
            ret += url[i];
        }
    }
    var data = ret.split('&');
    for(var i=0;i<data.length;i++)
    {
        var s = data[i].split("=");
        obj[s[0]] = s[1];
    }
    return obj;
}
